pkg/oauth/registry/oauthclientauthorization/etcd/etcd.go

```diff
@@ -19,27 +24,144 @@ type REST struct {
 	registry.Store
 }
 
-// NewREST returns a RESTStorage object that will work against oauth clients
-func NewREST(optsGetter restoptions.Getter, clientGetter oauthclient.Getter) (*REST, error) {
+type SelfREST struct {
+	helpers.ReadAndDeleteStorage
+}
+
+// NewREST returns a RESTStorage object that will work against oauth clients authorizations
+func NewREST(optsGetter restoptions.Getter, clientGetter oauthclient.Getter) (*REST, *SelfREST, error) {
+	resource, prefix, err := helpers.GetResourceAndPrefix(optsGetter, "oauthclientauthorizations")
+	if err != nil {
+		return nil, nil, fmt.Errorf("error building RESTOptions for %s store: %v", resource.String(), err)
+	}
+	strategy := oauthclientauthorization.NewStrategy(clientGetter)
 
 	store := &registry.Store{
 		NewFunc:     func() runtime.Object { return &api.OAuthClientAuthorization{} },
 		NewListFunc: func() runtime.Object { return &api.OAuthClientAuthorizationList{} },
+		KeyFunc: func(ctx kapi.Context, name string) (string, error) {
+			// Check to see if the name has the new format and thus is stored per user instead of a flat list
+			if username, clientname, err := helpers.SplitClientAuthorizationName(name); err == nil {
+				return registry.NoNamespaceKeyFunc(ctx, helpers.GetKeyWithUsername(prefix, username), clientname)
+			}
+			return registry.NoNamespaceKeyFunc(ctx, prefix, name)
+		},
 		ObjectNameFunc: func(obj runtime.Object) (string, error) {
 			return obj.(*api.OAuthClientAuthorization).Name, nil
 		},
-		PredicateFunc: func(label labels.Selector, field fields.Selector) *generic.SelectionPredicate {
-			return oauthclientauthorization.Matcher(label, field)
-		},
-		QualifiedResource: api.Resource("oauthclientauthorizations"),
+		PredicateFunc:     oauthclientauthorization.Matcher,
+		QualifiedResource: resource,
 
-		CreateStrategy: oauthclientauthorization.NewStrategy(clientGetter),
-		UpdateStrategy: oauthclientauthorization.NewStrategy(clientGetter),
+		CreateStrategy: strategy,
+		UpdateStrategy: strategy,
 	}
 
 	if err := restoptions.ApplyOptions(optsGetter, store, false, storage.NoTriggerPublisher); err != nil {
-		return nil, err
+		return nil, nil, err
+	}
+
+	selfStore := *store
+	selfStore.QualifiedResource = api.Resource("selfoauthclientauthorizations")
+	selfStore.CreateStrategy = helpers.CannotCreateStrategy // TODO maybe just set this to nil?
+	selfStore.UpdateStrategy = nil
+
+	// We cannot override KeyFunc because the cacher does not provide the user in the context
+	// The cacher does not use the KeyRootFunc so it is safe to override
+	selfStore.KeyRootFunc = func(ctx kapi.Context) string { // This makes watches more efficient
+		user, ok := kapi.UserFrom(ctx)
+		if !ok {
+			panic("User parameter required.")
```

Comment:
I don't think we should be panicing at this point, just return `kubeerr.NewBadRequest("User parameter required.")`, like below.